content/0007-standalone-workflow-execution/rfc.md

```diff
@@ -0,0 +1,294 @@
+if it does not exist.
+
+`relay cluster stop` - Stop all nodes in the cluster, sending proper termination
+signals to the Docker containers.
+
+`relay cluster delete` - Stop and delete the cluster and any metadata.
+
+`relay cluster kubectl` - Execute kubectl commands against the cluster.
+
+`relay cluster images import` - Imports an image from a remote or the local host into
+the cluster.
+
+`relay cluter images update` - Updates all images imported into the cluster.
+
+`relay config use-context (local|relay.sh)` - Updates the destination for relay
+commands.
+
+`relay config current-context` - Display the current context.
+
+#### Creating/Starting a cluster
+
+Starting and creating, for all intents and purposes, are the same thing. With
+`relay` you invoke the `cluster start` command to ensure a cluster is up and running,
+whether it exists or not.
+
+#### Stopping a cluster
+
+To stop a cluster, you invoke the `cluster stop` command and all the supporting
+containers will be shutdown.
+
+#### Destroying a cluster
+
+To destroy a cluster, you invoke the `cluster delete` command, which will stop
+and destroy running containers and delete any saved metadata and data
+files/directories. This can be used to start a new cluster from scratch.
+
+#### Using `kubectl`
+
+We don't want to mask the truth behind how workflows run. They run in Kubernetes
+as pods. Because Kubernetes is such a powerful system, we want to leverage it as
+much as possible. Especially when we are developing for the SaaS. If something
+fits better as a controller runtime than it does as a goroutine in the API, we
+want to encourage developers to try that. For this reason we need to expose the
+cluster to the user and ensure they can use standard kubectl commands against
+it.
+
+To ensure there's no confusion as to which cluster someone is accessing to
+perform maintenance, `kubectl` commands against the standalone cluster are
+invoked using `relay cluster kubectl`, which knows how to configure kubectl with
+the proper kubeconfig file.
+
+Obviously someone can choose to populate their global kubeconfig with the
+cluster credentials and change context if they want, but any mistakes caused
+doing so is on them.
+
+### Context switching
+
+Similar to kubectl, we add the ability to switch contexts to the Relay CLI. This
+allows us to setup a `local` context that will allow the relay command to run
+workflows against the local cluster. We add a default context called `relay.sh`
+that points to the production API. Any context that indicates a local
+standalone cluster will not require a login.
+
+#### Bootstrapping
+
+* `relay` creates a data directory in a known location
+* `relay` creates a json configuration file in the data directory
+* k3d creates a 3 node Kubernetes cluster and `relay` ensures it's running and
+  healthy
+* Critical 3rd party resources are deployed to the cluster using `kubectl apply`.
```

Comment:
Interestingly I think if the bootstrapping workflow doesn't need secrets, the only hard dependency is Tekton. The rest (Knative, Ambassador, Vault) can all be installed by a workflow!

Reply:
This is true, but I think it might make the process more complicated that way. I see secrets as a critical part of workflows and we wouldn't want to promote bad practices. It seems like if most workflows are going to contain some kind of secret, the default installation should include them right?